tool/resources/org/antlr/v4/tool/templates/codegen/Java/Java.stg

```diff
@@ -411,14 +411,19 @@ RuleFunction(currentRule,args,code,locals,ruleCtx,altLabelCtxs,namedActions,fina
 	<else>
 	catch (RecognitionException re) {
 		_localctx.exception = re;
-		_errHandler.reportError(this, re);
-		_errHandler.recover(this, re);
+		if (shouldReportAndRecover()) {
+			_errHandler.reportError(this, re);
+			_errHandler.recover(this, re);
+		}
 	}
 	<endif>
 	finally {
 		<finallyAction>
 		exitRule();
 	}
+	if (shouldReparseOrThrow(_localctx.exception)) {
```

Comment:
You are usually very concerned with class file size; won't this erase all your previous gains? Not saying it's wrong but... seems like something to consider.

Reply:
For me user convenience has more priority than generated/compiled size. Also it's not big increasing of genereted code (+5 small lines for every rule). If it's possible to decrease the code size than it should be decreased if it doesn't break other gains.

BTW, implemented architecture will allow SLL recover on real syntax error if such algorithm is implemented (I remember an issue about diambigution of real and errors because of SLL weakness).

Reply:
I updated Java.stg, now it's even less than 5 small lines :)